src/profiler.rs

```diff
@@ -465,6 +509,74 @@ impl Profiler {
     }
 }
 
+// The kernel thread ID ( the same returned by a call to gettid(2))
+// is not the same thing as the thread ID returned by pthread_self().
+//
+// Here we fetch a list of kernel thread IDs from /proc/{pid}/task
+#[cfg(target_os = "linux")]
+fn get_tids(pid_t: u32) -> Vec<u32> {
+    std::fs::read_dir(format!("/proc/{}/task", pid_t))
+        .unwrap()
+        .into_iter()
+        .filter_map(|entry| entry.map_or(None, |entry| entry.file_name().into_string().ok()))
+        .filter_map(|tid| tid.parse::<u32>().ok())
+        .collect()
+}
+
+#[cfg(target_os = "linux")]
+fn fire_rt_signals(running: Arc<AtomicBool>, frequency: u64) {
+    use std::{thread, time::Duration};
+
+    let pid = nix::unistd::Pid::this().as_raw() as u32;
+    let prof_tid = unsafe { libc::syscall(libc::SYS_gettid) as u32 };
+
+    let interval = 1e6 as u64 / frequency;
+    let tv_usec = interval % 1e6 as u64;
+
+    while running.load(Ordering::SeqCst) {
+        let tids = get_tids(pid);
+        for tid in tids {
+            // skip the thread running this function
+            if tid == prof_tid {
+                continue;
+            }
+            signal_thread(pid, tid, libc::SIGRTMIN());
```

Comment:
you should check the return value here and `continue` in case of an error. you don’t want to sleep if no thread was signalled (likely because it does not exist anymore?)

Reply:
@Swatinem this is a mistake. The thread is not supposed to sleep after signalling a single `tid`. The idea was to signal all the `tids` as quickly as possible, and then sleep for the given interval defined according to the frequency.

By reading your comment I've just realized I've mistakenly put the sleep inside the for loop instead of outside.